src/audio/mixin_mixout.c

```diff
@@ -262,113 +311,203 @@ static void audio_stream_bytes_copy(struct audio_stream __sparse_cache *dst_stre
+			n = MIN(audio_stream_bytes_without_wrap(source, src), n);
+			bytes_to_copy -= n;
+			n >>= 1;	/* divide by 2 */
+			for (i = 0; i < n; i++) {
+				int32_t s = (int32_t)*src;
+				s *= gain;
+				s >>= GAIN_DIV_SHIFT;
+				*dest = (int16_t)s;
+				src++;
+				dest++;
+			}
+
+			dest = audio_stream_wrap(sink, dest);
+			src = audio_stream_wrap(source, src);
+		}
+	}
 }
 #endif	/* CONFIG_FORMAT_S16LE */
 
 #if CONFIG_FORMAT_S24LE
 static void mix_s24(struct audio_stream __sparse_cache *sink, uint32_t start, uint32_t mixed_bytes,
-		    const struct audio_stream __sparse_cache *source, uint32_t size)
+		    const struct audio_stream __sparse_cache *source, uint32_t size, uint16_t gain)
 {
 	int32_t *dest, *src;
-	int bytes_processed;
+	uint32_t bytes_to_mix, bytes_to_copy;
 	int i, n;
 
 	dest = audio_stream_wrap(sink, (uint8_t *)sink->w_ptr + start);
 	src = source->r_ptr;
 
 	assert(mixed_bytes >= start);
-	uint32_t bytes_to_mix = MIN(mixed_bytes - start, size);
-
-	bytes_processed = 0;
+	bytes_to_mix = MIN(mixed_bytes - start, size);
+	bytes_to_copy = size - bytes_to_mix;
 
-	while (bytes_processed < bytes_to_mix) {
-		n = MIN(audio_stream_bytes_without_wrap(sink, dest),
-			bytes_to_mix - bytes_processed);
+	while (bytes_to_mix != 0) {
+		n = MIN(audio_stream_bytes_without_wrap(sink, dest), bytes_to_mix);
 		n = MIN(audio_stream_bytes_without_wrap(source, src), n);
+		bytes_to_mix -= n;
 		n >>= 2;	/* divide by 4 */
-		for (i = 0; i < n; i++) {
-			int32_t val1 = *dest << 8;	/* extending sign bit */
-			int32_t val2 = *src << 8;	/* extending sign bit */
-			*dest = sat_int24((val1 >> 8) + (val2 >> 8));
-			src++;
-			dest++;
+		if (gain == MIXIN_UNITY_GAIN) {
+			for (i = 0; i < n; i++) {
+				int32_t val1 = *dest << 8;	/* extending sign bit */
+				int32_t val2 = *src << 8;	/* extending sign bit */
+				*dest = sat_int24((val1 >> 8) + (val2 >> 8));
+				src++;
+				dest++;
+			}
+		} else {
+			for (i = 0; i < n; i++) {
+				int32_t d = *dest << 8;	/* extending sign bit */
+				int64_t s = (int64_t)(*src << 8);	/* extending sign bit */
+				s *= gain;
+				/* divide by 1024 and then shift back by 8 */
+				s >>= (GAIN_DIV_SHIFT + 8);
+				*dest = sat_int24((d >> 8) + (int32_t)s);
+				src++;
+				dest++;
+			}
 		}
 
-		bytes_processed += n << 2;
 		dest = audio_stream_wrap(sink, dest);
 		src = audio_stream_wrap(source, src);
 	}
 
-	audio_stream_bytes_copy(sink, dest, source, src, size - bytes_to_mix);
+	if (gain == MIXIN_UNITY_GAIN) {
+		audio_stream_bytes_copy(sink, dest, source, src, bytes_to_copy);
+	} else {
+		while (bytes_to_copy != 0) {
+			n = MIN(audio_stream_bytes_without_wrap(sink, dest), bytes_to_copy);
+			n = MIN(audio_stream_bytes_without_wrap(source, src), n);
+			bytes_to_copy -= n;
+			n >>= 2;	/* divide by 4 */
+			for (i = 0; i < n; i++) {
+				int64_t s = (int64_t)(*src << 8);	/* extending sign bit */
+				s *= gain;
+				s >>= (GAIN_DIV_SHIFT + 8);
+				*dest = (int32_t)s;
```

Comment:
sat_int24?

Reply:
Saturation check is not needed here as gain is always <=1.0 (1024 as integer). So the resulting value is equal or less then source value.